general/mem_manager.cpp

```diff
@@ -61,17 +83,424 @@ struct Ledger
+   { MFEM_CUDA_CHECK(::cudaMalloc(dptr, bytes)); }
+
+   void DeviceDealloc(void *dptr)
+   { MFEM_CUDA_CHECK(::cudaFree(dptr)); }
+
+   void *MemcpyHtoD(void *dst, const void *src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyHostToDevice));
+      return dst;
+   }
+
+   void *MemcpyDtoD(void* dst, const void* src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyDeviceToDevice));
+      return dst;
+   }
+
+   void *MemcpyDtoH(void *dst, const void *src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyDeviceToHost));
+      return dst;
+   }
+};
+
+/// The UVM device memory space. It is preferable to keep the cudaMemcpy
+/// in order to minimize the GPU page faults
+class UvmDeviceMemorySpace : public DeviceMemorySpace
+{
+public:
+   UvmDeviceMemorySpace(): DeviceMemorySpace() { }
+   void DeviceAlloc(void **dptr, const std::size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMallocManaged(dptr, bytes));
+   }
+   void DeviceDealloc(void *dptr)
+   {
+      MFEM_CUDA_CHECK(::cudaFree(dptr));
+   }
+   void *MemcpyHtoD(void *dst, const void *src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyHostToDevice));
+      return dst;
+   }
+
+   void *MemcpyDtoD(void* dst, const void* src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyDeviceToDevice));
+      return dst;
+   }
+
+   void *MemcpyDtoH(void *dst, const void *src, const size_t bytes)
+   {
+      MFEM_CUDA_CHECK(::cudaMemcpy(dst, src, bytes, cudaMemcpyDeviceToHost));
+      return dst;
+   }
+};
+
+#ifdef MFEM_USE_UMPIRE
+class UmpireDeviceMemorySpace : public DeviceMemorySpace
+{
+private:
+   umpire::ResourceManager& rm;
+   umpire::Allocator d_allocator;
+public:
+   UmpireDeviceMemorySpace():
+      DeviceMemorySpace(),
+      rm(umpire::ResourceManager::getInstance()),
+      d_allocator(rm.makeAllocator<umpire::strategy::DynamicPool>
+                  ("device_pool",rm.getAllocator("DEVICE"))) { }
+   void DeviceAlloc(void **dptr, const std::size_t bytes)
+   { *dptr = d_allocator.allocate(bytes); }
+   void DeviceDealloc(void *dptr)
+   { d_allocator.deallocate(dptr); }
+   void *MemcpyHtoD(void *dst, const void *src, const size_t bytes)
+   { rm.copy(dst, const_cast<void*>(src), bytes); return dst; }
+   void *MemcpyDtoD(void* dst, const void* src, const size_t bytes)
+   { rm.copy(dst, const_cast<void*>(src), bytes); return dst; }
+   void *MemcpyDtoH(void *dst, const void *src, const size_t bytes)
+   { rm.copy(dst, const_cast<void*>(src), bytes); return dst; }
+};
+#endif // MFEM_USE_UMPIRE 
+#endif // MFEM_USE_CUDA
+
+// *****************************************************************************
+
+/// (STD + NONE) memory space
+class StdNoneMemorySpace : public StdHostMemorySpace,
+   public NoneDeviceMemorySpace {};
```

Comment:
Defining a new class for each pair of Host and Device memory spaces seems like an overkill. Can't we just have one class with two variables inside?

Reply:
I updated the code with your approach: it's better that way, thanks!